### PR TITLE
Add script to compile release notes, update PR template and add tests

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

- Adds a script that compiles release notes based on merged PRs. It's also useful to see what happened between any two commits in the repo. You can try it like this:

  ```sh
  $ pip3 install GitPython
  $ python3 -m tools.CompileReleaseNotes
   ```
  
  It's particularly fancy if you pipe the output to a CLI Markdown renderer. E.g. install https://github.com/charmbracelet/glow and try this:

  ```sh
   $ python3 -m tools.CompileReleaseNotes | glow -
   ```
- Adds tests for this script in `tests/tools` and runs them on CI.
- Updates the PR template to reflect what information is pulled into the release notes. Removes the checkboxes for "bugfix" etc. because these can just be labels and I haven't seen them used in a useful way, most people don't tick them at all. As labels they are more visible and can be used in release notes.
- Since the release notes are compiled from PRs it would be good if we make sure they pass some basic checks. As an idea I added a CI test that checks the "code review checklist" below is ticked. I think that is a very quick thing to do and it makes sure the PR author has considered these items. I removed the "automated checks must pass" item from the checklist since that one is not necessary to check manually, it's built into the GitHub UI and checks can even be labeled "required". Ideally we would move more and more items from the checklist to automated tests.

### Code review checklist

- [x] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: tools/CompileReleaseNotes.py
Comment (nilsdeppe) [inline tools/CompileReleaseNotes.py]: should `no-github` and `github-token` be mutually exclusive?

File: tools/CompileReleaseNotes.py
Comment (nilsdeppe) [inline tools/CompileReleaseNotes.py]: Does the git module not have something similar to `git log  --first-parent`? This only shows merge commits in the history so you don't need to worry about matching things

File: .github/scripts/CheckPullRequest.py
Comment (nilsdeppe) [inline .github/scripts/CheckPullRequest.py]: [picking random line] if I understand this correctly, it adds a check (that we ideally would make as `Required`) that someone has gone and checked off all the elements in the PR review checklist? While I agree this is good in theory, I've yet to see a single PR where anything there gets checked. It's been difficult enough to get people to properly check of the type of change a PR makes, never mind anything more. I don't think we should add this check without bringing it up at the spectre call first. My worry is that this unfortunately may deter people from submitting PRs (GUIs can be a bit of a pain ;) )

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tools/CompileReleaseNotes.py
Comment (nilsvu) [reply tools/CompileReleaseNotes.py]: Yes, thanks for suggesting. I still have to do pattern matching to get the PR number though.

File: .github/scripts/CheckPullRequest.py
Comment (nilsvu) [reply .github/scripts/CheckPullRequest.py]: I agree, let's bring this up next Monday to hear what people think. Note that I created a new "PR automation" workflow for this test that runs on opening _and editing_ of PRs, not on pushes to the branch. So you can actually click the checkboxes above and after a few seconds the check updates, without running any of the other tests. You can test it by pushing this branch to your fork, open a PR to your fork's develop and click the checkboxes. This means the reviewers can just tick off the boxes before merging the PR, which also reminds _them_ to add appropriate labels to the PR if the author hasn't. That said, it could still turn out to be too annoying and we would be better off just not checking this and make it the responsibility of the person who merges the PR instead.

File: tools/CompileReleaseNotes.py
Comment (nilsdeppe) [inline tools/CompileReleaseNotes.py:255]: If I remember correctly there is some documentation that says this script requires `tqdm`. Since this appears to only be true if you aren't in `silent` mode, maybe update that when you squash?

**Review by nilsdeppe — COMMENTED** [review]
Looks good, please squash

File: tools/CompileReleaseNotes.py
Comment (nilsvu) [reply tools/CompileReleaseNotes.py:255]: That's in the dev guide I add in https://github.com/sxs-collaboration/spectre/pull/2598 :) Will update it there.

Comment (nilsdeppe) [reply tools/CompileReleaseNotes.py:255]: :+1:

**Comment by nilsvu** [comment]
Following the discussion at the meeting today I dropped the CI workflow that tests the checklist in a PR is ticked off because it would probably be too annoying.

Only the commit "Update PR template" changed. @nilsdeppe could you take another look at that commit? @wthrowe would you do the second review?

**Review by nilsdeppe — DISMISSED** [review]
@wthrowe would you mind looking at this? It got factored out of #2598, which you looked at.

File: tests/tools/Test_CompileReleaseNotes.py
Comment (wthrowe) [inline tests/tools/Test_CompileReleaseNotes.py:146]: Why are these linked sometimes, but not all of the time?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wthrowe) [inline .github/PULL_REQUEST_TEMPLATE.md]: Many people don't know what an HTML comment looks like.  Maybe "between the two lines", or, to be explicit, "between the two UPGRADE INSTRUCTIONS lines".

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/tools/Test_CompileReleaseNotes.py
Comment (nilsvu) [reply tests/tools/Test_CompileReleaseNotes.py:146]: Because I added the `url` field to only some of the test-objects. If no `url` is available then the ID is not linked. In practice, URLs will be available when GitHub is queried, and won't be available when GitHub-queries are disabled with the `--no-github` flag.

**Comment by nilsvu** [comment]
@wthrowe reworded the PR template as you suggested, could you take another look?

**Review by wthrowe — APPROVED** [review]
Looks good.

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
